src/metrics.jl

```diff
@@ -99,211 +182,163 @@ struct MeanSqDeviation <: SemiMetric end
+    evaluate(d, a, b, parameters(d))
+end
 
-# Specialized for Arrays and avoids a branch on the size
-@inline Base.@propagate_inbounds function evaluate(d::UnionMetrics, a::Union{Array, ArraySlice}, b::Union{Array, ArraySlice})
+@inline Base.@propagate_inbounds function evaluate(d::UnionMetrics, a::AbstractArray, b::AbstractArray, ::Nothing)
     @boundscheck if length(a) != length(b)
         throw(DimensionMismatch("first array has length $(length(a)) which does not match the length of the second, $(length(b))."))
     end
-    p = parameters(d)
-    @boundscheck if p !== nothing
-        length(a) != length(p) && throw(DimensionMismatch("arrays have length $(length(a)) but parameters have length $(length(p))."))
-    end
     if length(a) == 0
         return zero(result_type(d, a, b))
     end
     @inbounds begin
         s = eval_start(d, a, b)
-        if p === nothing
+        if IndexStyle(a, b) === IndexLinear()
             @simd for I in 1:length(a)
                 ai = a[I]
                 bi = b[I]
                 s = eval_reduce(d, s, eval_op(d, ai, bi))
             end
         else
-            @simd for I in 1:length(a)
-                aI = a[I]
-                bI = b[I]
-                pI = p[I]
-                s = eval_reduce(d, s, eval_op(d, aI, bI, pI))
+            if size(a) == size(b)
+                @simd for I in eachindex(a, b)
+                    ai = a[I]
+                    bi = b[I]
+                    s = eval_reduce(d, s, eval_op(d, ai, bi))
+                end
+            else
+                for (Ia, Ib) in zip(eachindex(a), eachindex(b))
+                    ai = a[Ia]
+                    bi = b[Ib]
+                    s = eval_reduce(d, s, eval_op(d, ai, bi))
+                end
             end
         end
         return eval_end(d, s)
     end
 end
 
-@inline function evaluate(d::UnionMetrics, a::AbstractArray, b::AbstractArray)
+@inline Base.@propagate_inbounds function evaluate(d::UnionMetrics, a::AbstractArray, b::AbstractArray, p::AbstractArray)
     @boundscheck if length(a) != length(b)
         throw(DimensionMismatch("first array has length $(length(a)) which does not match the length of the second, $(length(b))."))
     end
-    p = parameters(d)
-    @boundscheck if p !== nothing
-        length(a) != length(p) && throw(DimensionMismatch("arrays have length $(length(a)) but parameters have length $(length(p))."))
+    @boundscheck if length(a) != length(p)
+        throw(DimensionMismatch("arrays have length $(length(a)) but parameters have length $(length(p))."))
     end
     if length(a) == 0
         return zero(result_type(d, a, b))
     end
     @inbounds begin
         s = eval_start(d, a, b)
-        if size(a) == size(b)
-            if p === nothing
-                @simd for I in eachindex(a, b)
+        if IndexStyle(a, b, p) === IndexLinear()
+            @simd for I in 1:length(a)
+                ai = a[I]
+                bi = b[I]
+                pi = p[I]
```

Comment:
Is there any issue with overwriting the Base constant `pi`? I was experimenting with that and wasn't sure, so I used `pI` and `aI` etc.

Reply:
I don't think so. Since we don't want to access the mathematical constant here, it should not matter that it is masked by the local variable.